pkg/util/arm/arm.go

```diff
@@ -1,22 +1,264 @@
+// on its types.  It then merges remaining fields of r over the result
+func (r *Resource) MarshalJSON() ([]byte, error) {
+	// Make a shadow copy of the `r.Resource`
+	resource := shadowCopy(r.Resource)
+
+	resourceV := reflect.ValueOf(resource)
+	outerV := reflect.ValueOf(*r)
+
+	// Check that the underlying resource is a struct
+	if resourceV.Kind() != reflect.Struct {
+		return nil, errors.New("Resource field of the arm.Resource must be a struct")
+	}
+
+	// Create a list of fields that combines fields of `r.Resource` and outer `*r` structs.
+	// Fields from the outer struct `*r` override fields from `r.Resource`.
+	fields := make([]reflect.StructField, 0, resourceV.NumField())
+	fieldsMap := map[string]int{}
+	// Copy fields from `r.Resource`
+	for i := 0; i < resourceV.NumField(); i++ {
+		field := resourceV.Type().Field(i)
+		fields = append(fields, field)
+		fieldsMap[field.Name] = i
+	}
+	// Copy fields from `*r` & override if already exists
+	for i := 0; i < outerV.NumField(); i++ {
+		outerField := outerV.Type().Field(i)
+		if outerField.Name == "Resource" {
+			continue
+		}
+		if fieldsIndex, found := fieldsMap[outerField.Name]; found {
+			outerField.Type = emptyInterfaceType
+			fields[fieldsIndex] = outerField
+		} else {
+			fields = append(fields, outerField)
+		}
+	}
+
+	// Create a new struct of fields
+	combinedV := reflect.New(reflect.StructOf(fields)).Elem()
+	// Copy values from `r.Resource`
+	for i := 0; i < resourceV.NumField(); i++ {
+		field := resourceV.Type().Field(i)
+		fieldV := resourceV.Field(i)
+		combinedV.FieldByName(field.Name).Set(fieldV)
+	}
+	// Copy values from `*r`
+	for i := 0; i < outerV.NumField(); i++ {
+		field := outerV.Type().Field(i)
+		fieldV := outerV.Field(i)
+		if field.Name == "Resource" {
+			continue
+		}
+		if !isZero(fieldV) {
+			combinedV.FieldByName(field.Name).Set(fieldV)
+		}
+	}
+
+	return json.Marshal(combinedV.Interface())
+}
+
+// UnmarshalJSON is not implemented
+func (r *Resource) UnmarshalJSON(b []byte) error {
+	return fmt.Errorf("not implemented")
+}
+
+var (
+	stringType         = reflect.TypeOf("")
+	emptyInterfaceType = reflect.ValueOf([]interface{}(nil)).Type().Elem()
+)
+
+// shadowCopy returns a copy of the input object wherein all the struct types
+// have been replaced with dynamically created ones.  The idea is that the
+// JSONMarshal() methods get dropped in the process and so the returned object
+// marshals natively.  Type cycles are permitted, but (as in encoding/json)
+// value cycles are not.  Golang reflect doesn't support dynamically creating
+// named types; to get around this we go weakly typed
+func shadowCopy(i interface{}) interface{} {
+	return _shadowCopy(reflect.ValueOf(i)).Interface()
+}
+
+func _shadowCopy(v reflect.Value) reflect.Value {
+	switch v.Kind() {
+	case reflect.Array:
+		a := reflect.New(reflect.ArrayOf(v.Len(), emptyInterfaceType)).Elem()
+		for i := 0; i < v.Len(); i++ {
```

Comment:
v4 needed a fix for the uuid.UUID - that would probably be good to backport as a separate PR to keep in sync